core/blockchain.go

```diff
@@ -2567,14 +2586,21 @@ func (bc *BlockChain) writeValidatorSnapshots(addrs []common.Address) error {
 		}
 		validators = append(validators, validator)
 	}
-
+	epoch := bc.CurrentHeader().Epoch()
 	// Batch write the current data as snapshot
 	batch := bc.db.NewBatch()
 	for i := range validators {
-		err := rawdb.WriteValidatorSnapshot(batch, validators[i])
+		stats, err := rawdb.ReadValidatorStats(bc.db, validators[i].Address)
 		if err != nil {
 			return err
 		}
+		validators[i].Snapshot.Epoch = epoch
+		validators[i].Snapshot.NumBlocksSigned = stats.NumBlocksSigned
```

Comment:
Remove the NumBlocksSigned and ToSign in ValidatorStats and only keep those in Wrapper as source of truth. And update those number in Wrapper every block. Then here the snapshot process will be simply copy and store the wrapper as is.